br/pkg/restore/import_retry.go

```diff
@@ -0,0 +1,248 @@
+	start      []byte
+	end        []byte
+	metaClient SplitClient
+
+	errors error
+	rs     utils.RetryState
+}
+
+// OverRegionsInRange creates a controller that cloud be used to scan regions in a range and
+// apply a function over these regions.
+// You can then call the `Run` method for applying some functions.
+func OverRegionsInRange(start, end []byte, metaClient SplitClient, retryStatus utils.RetryState) OverRegionsInRangeController {
+	return OverRegionsInRangeController{
+		start:      start,
+		end:        end,
+		metaClient: metaClient,
+		rs:         retryStatus,
+	}
+}
+
+func (o *OverRegionsInRangeController) onError(ctx context.Context, result RPCResult, region *RegionInfo) {
+	o.errors = multierr.Append(o.errors, errors.Annotatef(&result, "execute over region %v failed", region.Region))
+	// TODO: Maybe handle some of region errors like `epoch not match`?
+}
+
+func (o *OverRegionsInRangeController) tryFindLeader(ctx context.Context, region *RegionInfo) (*metapb.Peer, error) {
+	var leader *metapb.Peer
+	failed := false
+	leaderRs := utils.InitialRetryState(4, 5*time.Second, 10*time.Second)
+	err := utils.WithRetry(ctx, func() error {
+		r, err := o.metaClient.GetRegionByID(ctx, region.Region.Id)
+		if err != nil {
+			return err
+		}
+		if !checkRegionEpoch(r, region) {
+			failed = true
+			return nil
+		}
+		if r.Leader != nil {
+			leader = r.Leader
+			return nil
+		}
+		return errors.Annotatef(berrors.ErrPDLeaderNotFound, "there is no leader for region %d", region.Region.Id)
+	}, &leaderRs)
+	if failed {
+		return nil, errors.Annotatef(berrors.ErrKVEpochNotMatch, "the current epoch of %s is changed", region)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return leader, nil
+}
+
+// handleInRegionError handles the error happens internal in the region. Update the region info, and perform a suitable backoff.
+func (o *OverRegionsInRangeController) handleInRegionError(ctx context.Context, result RPCResult, region *RegionInfo) (cont bool) {
+
+	if nl := result.StoreError.GetNotLeader(); nl != nil {
+		if nl.Leader != nil {
+			region.Leader = nl.Leader
+			// try the new leader immediately.
+			return true
+		}
+		// we retry manually, simply record the retry event.
+		o.rs.RecordRetry()
+		// There may not be leader, waiting...
+		leader, err := o.tryFindLeader(ctx, region)
+		if err != nil {
+			// Leave the region info unchanged, let it retry then.
+			logutil.CL(ctx).Warn("failed to find leader", logutil.Region(region.Region), logutil.ShortError(err))
+			return false
+		}
+		region.Leader = leader
+		return true
+	}
+	// For other errors, like `ServerIsBusy`, `RegionIsNotInitialized`, just trivially backoff.
+	time.Sleep(o.rs.ExponentialBackoff())
+	return true
+}
+
+// Run executes the `regionFunc` over the regions in `o.start` and `o.end`.
+// It would retry the errors according to the `rpcResponse`.
+func (o *OverRegionsInRangeController) Run(ctx context.Context, f RegionFunc) error {
+	if !o.rs.ShouldRetry() {
```

Comment:
retryStatus works on some regions。
Retry time(16) is small for a mount of regions ?